### ISSUE TITLE
Dtype of reverse-mode gradient is incorrect when input dtype is a mixture of float32 and complex128

### ISSUE THREAD

**Issue by eeDigitalSeal** [tensorflow/tensorflow#56982]
### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.9

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.9

### Current Behaviour?

Like the case below, reverse mode AD gives the gradient in float32 dtype when the input dtype is a mixture of float32 and complex128. It should be complex128.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

x = tf.constant(
[[0.69053245, 0.7255454, 0.4136572 ],
 [0.6055007, 0.5583403, 0.07835412]], shape=(2, 3), dtype=tf.float32)

y = tf.constant(
[[0.30684007+0.04178488j, 0.0126979 +0.32119822j, 0.99649133+0.13754789j],
 [0.29316908+0.50216602j, 0.11820176+0.79961702j, 0.41341467+0.78824452j]], shape=(2, 3), dtype=tf.complex128)

with tf.GradientTape(persistent=True) as g:
 g.watch(x)
 g.watch(y)
 res_backward = tf.keras.losses.MSE(x, y)
grad_backward = g.gradient(res_backward,x)
print(grad_backward)

grad_fwd_arr = []

for i in range(tf.size(x)):
    tangents = tf.reshape(tf.one_hot(i,tf.size(x)),shape=x.shape)
    with tf.autodiff.ForwardAccumulator(x, tangents) as acc:
        res_forward = tf.keras.losses.MSE(x,y)
        jvp = acc.jvp(res_forward)
        grad_fwd_arr.append(tf.reduce_sum(jvp))

grad_fwd = tf.reshape(tf.convert_to_tensor(grad_fwd_arr),shape=x.shape)
print(grad_fwd)
```

### Relevant log output

```shell
tf.Tensor(
[[ 0.2557949   0.47523168 -0.3885561 ]
 [ 0.20822108  0.2934257  -0.2233737 ]], shape=(2, 3), dtype=float32)
tf.Tensor(
[[ 0.25579492+0.02785659j  0.47523167+0.21413215j -0.38855609+0.09169859j]
 [ 0.20822108+0.33477735j  0.2934257 +0.53307801j -0.2233737 +0.52549635j]], shape=(2, 3), dtype=complex128)
```

**Comment by sushreebarsa** [comment]
@gadagashwini I was able to replicate the issue, please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/eb1283d233d921d690e7401dd96db5be/56982.ipynb) for reference. Thank you!

**Comment by sachinprasadhs** [comment]
`tf.keras.losses.MSE` is doing the casting and it is converting to complex128 in both the cases, but in `grad_backward` when you calculate the gradient with respect to x, which is float32, it casts back to float32, that is the reason you're getting float32 in `grad_backward` and `complex128` in `res_forward`. 
To maintain the consistency, you can include manual casting to the `dtype` which you require.
